lib/charms/opensearch/v0/opensearch_nodes_exclusions.py

```diff
@@ -0,0 +1,163 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Base class for OpenSearch node exclusions management."""
+import logging
+from functools import cached_property
+from typing import List, Optional, Set
+
+from charms.opensearch.v0.helper_databag import Scope
+from charms.opensearch.v0.models import Node
+from charms.opensearch.v0.opensearch_exceptions import (
+    OpenSearchError,
+    OpenSearchHttpError,
+)
+
+# The unique Charmhub library identifier, never change it
+LIBID = "51c1ac864e9a4d12b1d1ef27c0ff2e50"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+logger = logging.getLogger(__name__)
+
+
+class OpenSearchExclusionsException(OpenSearchError):
+    """Exception class for all Voting/Allocation exclusions related exceptions."""
+
+
+ALLOCS_TO_DELETE = "allocation-exclusions-to-delete"
+VOTING_TO_DELETE = "delete-voting-exclusions"
+
+
+class OpenSearchExclusions:
+    """Exclusions related operations, important to reason as exclusions, NOT additions."""
+
+    def __init__(self, charm):
+        self._charm = charm
+        self._opensearch = self._charm.opensearch
+
+        self._scope = Scope.APP if self._charm.unit.is_leader() else Scope.UNIT
+
+    def add_current(self) -> None:
+        """Add Voting and alloc exclusions."""
+        if (self._node.is_cm_eligible() or self._node.is_voting_only()) and not self._add_voting():
+            logger.error(f"Failed to add voting exclusion: {self._node.name}.")
+
+        if self._node.is_data() and not self._add_allocations():
+            logger.error(f"Failed to add shard allocation exclusion: {self._node.name}.")
+
+    def delete_current(self) -> None:
+        """Delete Voting and alloc exclusions."""
+        if (
+            self._node.is_cm_eligible() or self._node.is_voting_only()
+        ) and not self._delete_voting():
+            self._charm.peers_data.put(self._scope, VOTING_TO_DELETE, True)
+
+        if self._node.is_data() and not self._delete_allocations():
+            current_allocations = self._charm.peers_data.get(self._scope, ALLOCS_TO_DELETE)
+            if not current_allocations:
+                return
+
+            current_allocations = set(current_allocations.split(","))
+            if self._node.name in current_allocations:
+                current_allocations.remove(self._node.name)
+                self._charm.peers_data.put(
+                    self._scope, ALLOCS_TO_DELETE, ",".join(current_allocations)
+                )
```

Comment:
Why did we remove the node from the ALLOCS_TO_DELETE list if we were NOT able to delete it from OpenSearch?

Reply:
Good catch, that was wrong